examples/wakaama/lwm2mclient.c

```diff
@@ -0,0 +1,625 @@
+            break;
+         case STATE_BOOTSTRAPPING:
+            fprintf(stdout, "STATE_BOOTSTRAPPING\r\n");
+            break;
+         case STATE_REGISTER_REQUIRED:
+            fprintf(stdout, "STATE_REGISTER_REQUIRED\r\n");
+            break;
+         case STATE_REGISTERING:
+            fprintf(stdout, "STATE_REGISTERING\r\n");
+            break;
+         case STATE_READY:
+            fprintf(stdout, "STATE_READY\ren");
+            break;
+         default:
+            fprintf(stdout, "Unknown...\r\n");
+            break;
+      }
+      if (result != 0) {
+         fprintf(stderr, "lwm2m_step() failed: 0x%X\r\n", result);
+#ifdef LWM2M_BOOTSTRAP
+         if (previousState == STATE_BOOTSTRAPPING) {
+#ifdef WITH_LOGS
+            fprintf(stdout,
+                    "[BOOTSTRAP] restore security and server objects\r\n");
+#endif
+            prv_restore_objects(lwm2mH);
+            lwm2mH->state = STATE_INITIAL;
+         } else
+#endif
+            return NULL;
+      }
+#ifdef LWM2M_BOOTSTRAP
+      DEBUG("update_bootstrap_info()\r\n");
+      update_bootstrap_info(&previousState, lwm2mH);
+#endif
+
+   handshake:
+      DEBUG("xtimer_msg_receive_timeout64()\r\n");
+      //if (xtimer_msg_receive_timeout64(&msg, tv.tv_sec * 1000000) >= 0) {
+      pckt_rcvd_size = sock_udp_recv(&data.sock, &packet_rcvd, DTLS_MAX_BUF, tv.tv_sec * US_PER_SEC, &remote);
+      if (pckt_rcvd_size > 0) {
+         connP = connection_find(data.connList, &remote);
+         if (connP != NULL) {
+            DEBUG("connection_handle_packet(%d)\r\n", pckt_rcvd_size);
+            int result = connection_handle_packet(connP, packet_rcvd, pckt_rcvd_size);
+            if (0 != result) {
+               printf("error handling message %d\n", result);
+            }
+
+            /*
+             * We loop around here until the DTLS handshake is done,
+             * since lwm2m_step() will try to send stuff before then and fail
+             * without retrying.
+             */
+            if (!connP->noSec) {
+               dtls_peer_t *peer =
+                   dtls_get_peer(connP->dtlsContext, connP->dtlsSession);
+               if (peer->state != DTLS_STATE_CONNECTED) {
+                  goto handshake;
+               }
+            }
+         }
+      }
+      // TODO: Error handling for sock_udp_recv()
+   }
+
+   /*
+    * Finally when the loop is left smoothly - asked by user in the command line
+    * interface - we unregister our client from it
+    */
+   if (g_quit == 1) {
+      lwm2m_free(pskBuffer);
+
+#ifdef LWM2M_BOOTSTRAP
+      close_backup_object();
+#endif
+      lwm2m_close(lwm2mH);
+   }
+   //close(data.sock);
+   sock_udp_close(&data.sock);
+   connection_free(data.connList);
+
+   clean_security_object(objArray[0]);
+   lwm2m_free(objArray[0]);
+   clean_server_object(objArray[1]);
+   lwm2m_free(objArray[1]);
+   free_object_device(objArray[2]);
+   acl_ctrl_free_object(objArray[3]);
+
+   return NULL;
```

Comment:
Resources are not being released correctly, I'm forced to kill the instance native to be able to quit.

Reply:
Oh, so that's why it doesn't terminate with ctrl+c on native? I was wondering why that is.

I've replaced the memarray stuff in platform.c with malloc/free again and looked what valgrind says. I did a small fix, but there seems to be a leak from TinyDTLS. Not sure if that's it or what to do about it.

Reply:
It was not memarray. This issue is present in the bb8ca70d4d33212d3a7b37b57fc1c25a2f38160f commit.

Reply:
I know, it's been there from the start. I only disabled memarray so valgrind would also detect unfreed memory in the memarray stuff (not that it would matter, since all that is statically allocated with memarray, now that I think about it).

Anyway, the client-thread seems to terminate. At least `ps` doesn't show it anymore after a ctrl+c. I seem to have forgotten to close the socks allocated for the individual connection, but since I still can't get a network connection under native, none ever gets created anyway, aside from the listener.

Not sure how to debug this further/what else to look at.

Reply:
I still have no idea which blocking resources are not cleaned up properly. For the time being, I added a workaround so the programm can be closed on native without having to kill the process (exit command in the shell).